rllib/agents/alpha_star/alpha_star.py

```diff
@@ -0,0 +1,592 @@
+            "policies": {
+                # Our main policy, we'd like to optimize.
+                "main_0": PolicySpec(),
+                # Initial main exploiters (1 of these is a random policy).
+                "main_exploiter_0": PolicySpec(policy_class=RandomPolicy),
+                "main_exploiter_1": PolicySpec(),
+                "main_exploiter_2": PolicySpec(),
+                "main_exploiter_3": PolicySpec(),
+                "main_exploiter_4": PolicySpec(),
+                # Initial league exploiter (1 of these is a random policy).
+                "league_exploiter_0": PolicySpec(policy_class=RandomPolicy),
+                "league_exploiter_1": PolicySpec(),
+                "league_exploiter_2": PolicySpec(),
+                "league_exploiter_3": PolicySpec(),
+                "league_exploiter_4": PolicySpec(),
+            },
+            "policy_mapping_fn":
+                (lambda aid, ep, worker, **kw: "main_0" if
+                 ep.episode_id % 2 == aid else "main_exploiter_0"),
+            # Train all non-random policies that exist at beginning.
+            "policies_to_train": [
+                "main_0", "main_exploiter_1", "main_exploiter_2",
+                "main_exploiter_3", "main_exploiter_4", "league_exploiter_1",
+                "league_exploiter_2", "league_exploiter_3",
+                "league_exploiter_4",
+            ],
+        },
+
+        # Reporting interval.
+        "min_time_s_per_reporting": 2,
+
+        # Use the `training_iteration` method instead of an execution plan.
+        "_disable_execution_plan_api": True,
+    },
+    _allow_unknown_configs=True,
+)
+
+# __sphinx_doc_end__
+# yapf: enable
+
+
+class AlphaStarTrainer(appo.APPOTrainer):
+    @classmethod
+    @override(Trainer)
+    def default_resource_request(cls, config):
+        cf = dict(cls.get_default_config(), **config)
+
+        num_policies = len(cf["max_num_policies_to_train"])
+        if cf["num_gpus"]:
+            num_learner_shards = min(cf["num_gpus"], num_policies)
+            num_gpus_per_shard = cf["num_gpus"] / num_learner_shards
+        else:
+            num_learner_shards = cf.get("num_replay_buffer_shards", 1)
+            num_gpus_per_shard = 0
+
+        eval_config = cf["evaluation_config"]
+
+        # Return PlacementGroupFactory containing all needed resources
+        # (already properly defined as device bundles).
+        return PlacementGroupFactory(
+            bundles=[{
+                # Driver (no GPUs).
+                "CPU": cf["num_cpus_for_driver"],
+            }] + [
+                {
+                    # RolloutWorkers (no GPUs).
+                    "CPU": cf["num_cpus_per_worker"],
+                } for _ in range(cf["num_workers"])
+            ] + [
+                {
+                    # Policy learners (and Replay buffer shards).
+                    "CPU": 1,
+                    "GPU": num_gpus_per_shard,
```

Comment:
actually what happens if num_gpus_per_shard is a fload num?
will core handle this correctly?
should we always round up?

Reply:
No, this will be fine. You can give fractions for GPUs.

Reply:
Rounding down now, but only for > 1.0.
Otherwise, you'd over-allocate (use more GPUs than we have).